File: 300.500/436.FindRightInterval.cpp
#include <iostream>
#include <vector>
#include <unordered_map>
#include <algorithm>
using namespace std;

class FindRightInterval
{
public:
    vector<int> findRightInterval(vector<vector<int>>& intervals) {
        //记录原始下标
        kv.clear();
        for (int i = 0; i < intervals.size(); ++i) {
            kv[intervals[i][0]] = i;
        }
        //排序
        sort(intervals.begin(), intervals.end(), [](const vector<int>& l, const vector<int>& r)->bool {
            return l[0] < r[0];
            });
        //筛选
        vector<int> ret(intervals.size());
        for (int i = 0;i < intervals.size();++i) {
            int k = kv[intervals[i][0]];
            ret[k] = rightInterval(intervals, i);
        }
        return ret;
    }
private:
    int rightInterval(vector<vector<int>>& intervals, int start) {
        for (int j = start;j < intervals.size();++j) {
            if (intervals[j][0] >= intervals[start][1]) {
                return kv[intervals[j][0]];
            }
        }
        return -1;
    }
private:
    unordered_map<int, int> kv;
public:
    vector<int> findRightInterval1(vector<vector<int>>& intervals) {
        int n = intervals.size();
        vector<pair<int,int>> dp(n);
        for (int i = 0; i < n; ++i) {
            dp[i] = make_pair(intervals[i][0], i);
        }
        sort(dp.begin(), dp.end());
        vector<int> ret(n, -1);
        for (int i = 0; i < n;++i) {
            auto it = lower_bound(dp.begin(), dp.end(), make_pair(intervals[i][1], 0));
            if (it != dp.end()) {
                ret[i] = it->second;
            }
        }
        return ret;
    }
};

void Print(vector<int>& ret) {
    cout << "[";
    for (auto x : ret) {
        cout << x << " ,";
    }
    cout << "]" << endl;
}

int main(int argc, char const* argv[]) {
    FindRightInterval cls;
    auto& ret1 = cls.findRightInterval1(vector<vector<int>>{ {1, 2}});
    Print(ret1);
    auto& ret2 = cls.findRightInterval1(vector<vector<int>>{ {3, 4}, { 2,3 }, { 1, 2 }});
    Print(ret2);
    auto& ret3 = cls.findRightInterval1(vector<vector<int>>{ {1, 4}, { 2,3 }, { 3,4 }});
    Print(ret3);
    return 0;
}

File: go/300.500/436.FindRightInterval.go
package main

import (
	"fmt"
	"sort"
)

func findRightInterval(intervals [][]int) []int {
	n := len(intervals)
	kv := make(map[int]int, n)
	for i, x := range intervals {
		kv[x[0]] = i
	}
	sort.Slice(intervals, func(i, j int) bool {
		return intervals[i][0] < intervals[j][0]
	})
	fn := func(start int) int {
		for i := start; i < n; i++ {
			if intervals[i][0] >= intervals[start][1] {
				return kv[intervals[i][0]]
			}
		}
		return -1
	}
	ret := make([]int, n)
	for i, x := range intervals {
		k := kv[x[0]]
		ret[k] = fn(i)
	}
	return ret
}

func findRightInterval1(intervals [][]int) []int {
	n := len(intervals)
	dp := make([][]int, n)
	for i, x := range intervals {
		dp[i] = []int{x[0], i}
	}
	sort.Slice(dp, func(i, j int) bool {
		return dp[i][0] < dp[j][0]
	})
	ret := make([]int, n)
	for i, x := range intervals {
		j := sort.Search(n, func(j int) bool {
			return dp[j][0] >= x[1]
		})
		if j < n {
			ret[i] = dp[j][1]
		} else {
			ret[i] = -1
		}
	}
	return ret
}

func findRightInterval2(intervals [][]int) []int {
	for i := range intervals {
		intervals[i] = append(intervals[i], i)
	}
	n := len(intervals)
	sort.Slice(intervals, func(i, j int) bool {
		return intervals[i][0] < intervals[j][0]
	})
	ret := make([]int, n)
	for _, x := range intervals {
		j := sort.Search(n, func(j int) bool {
			return intervals[j][0] >= x[1]
		})
		if j < n {
			ret[x[2]] = intervals[j][2]
		} else {
			ret[x[2]] = -1
		}
	}
	return ret
}

func main() {
	fmt.Println(findRightInterval2([][]int{{1, 2}}))
	fmt.Println(findRightInterval2([][]int{{3, 4}, {2, 3}, {1, 2}}))
	fmt.Println(findRightInterval2([][]int{{1, 4}, {2, 3}, {3, 4}}))
	fmt.Println(findRightInterval2([][]int{{1, 1}, {3, 4}}))
}
